arangod/Cluster/ClusterMethods.cpp

```diff
@@ -1920,90 +1977,91 @@ int fetchEdgesFromEngines(std::string const& dbname,
 ///        a 'null' will be inserted into the result.
 
 void fetchVerticesFromEngines(
-    std::string const& dbname,
+    transaction::Methods& trx,
     std::unordered_map<ServerID, traverser::TraverserEngineID> const* engines,
     std::unordered_set<arangodb::velocypack::StringRef>& vertexIds,
-    std::unordered_map<arangodb::velocypack::StringRef, std::shared_ptr<VPackBuffer<uint8_t>>>& result,
-    VPackBuilder& builder) {
-  auto cc = ClusterComm::instance();
-  if (cc == nullptr) {
-    // nullptr happens only during controlled shutdown
-    return;
-  }
+    std::unordered_map<arangodb::velocypack::StringRef, std::shared_ptr<VPackBuffer<uint8_t>>>& result) {
   // TODO map id => ServerID if possible
   // And go fast-path
 
   // slow path, sharding not deducable from _id
-  builder.clear();
-  builder.openObject();
-  builder.add(VPackValue("keys"));
-  builder.openArray();
+  transaction::BuilderLeaser lease(&trx);
+  lease->clear();
+  lease->openObject();
+  lease->add(VPackValue("keys"));
+  lease->openArray();
   for (auto const& v : vertexIds) {
     // TRI_ASSERT(v.isString());
-    builder.add(VPackValuePair(v.data(), v.length(), VPackValueType::String));
+    lease->add(VPackValuePair(v.data(), v.length(), VPackValueType::String));
   }
-  builder.close();  // 'keys' Array
-  builder.close();  // base object
+  lease->close();  // 'keys' Array
+  lease->close();  // base object
 
-  std::string const url =
-      "/_db/" + StringUtils::urlEncode(dbname) + "/_internal/traverser/vertex/";
+  std::string const url = "/_db/" + StringUtils::urlEncode(trx.vocbase().name()) +
+                          "/_internal/traverser/vertex/";
 
-  std::vector<ClusterCommRequest> requests;
-  auto body = std::make_shared<std::string>(builder.toJson());
+  auto* pool = trx.vocbase().server().getFeature<NetworkFeature>().pool();
+
+  std::vector<Future<network::Response>> futures;
+  futures.reserve(engines->size());
+
+  VPackBufferUInt8 buffer;
+  buffer.append(lease->data(), lease->size());
```

Comment:
Why is it necessary to copy everything from `lease` into `buffer` here?
`buffer` is copied when we call `network::sendRequestRetry` and not modified, so we could instead pass the already existing velocypack::Buffer of `lease` into `network::sendRequestTry`, and get rid of `buffer` entirely.